### PR TITLE
ARROW-13629: [Ruby] Add support for building/converting map

### PR DESCRIPTION
**Pull request by kanga333**

### CODE REVIEW

**Comment by kanga333** [comment]
[TODO] improve test cases.

File: ruby/red-arrow/lib/arrow/map-array-builder.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    #
+    #   Appends a map record including key and item values.
+    #
+    #   @param value [nil, #each] The map record.
+    #
+    #     If this is `nil`, the map record is null.
+    #
+    #     If this is an `Object` that has `#each`, each value is a pair of key and item.
+    #
+    # @since 5.0.0
```
Comment (kou) [inline ruby/red-arrow/lib/arrow/map-array-builder.rb:44]:
```suggestion
    # @since 6.0.0
```

File: ruby/red-arrow/test/test-map-data-type.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    test("ordered arguments") do
+      assert_equal("map<int8, string>",
+                   Arrow::MapDataType.new(@key, @item).to_s)
+    end
+
+    test("description") do
+      assert_equal("map<int8, string>",
+                   Arrow::MapDataType.new(key: @key,
+                                          item: @item,).to_s)
```
Comment (kou) [inline ruby/red-arrow/test/test-map-data-type.rb:33]:
```suggestion
                                          item: @item).to_s)
```

File: ruby/red-arrow/test/values/test-sparse-union-array.rb [hunk 3]
```diff
@@ -373,6 +373,20 @@ def test_struct
...
+      {"0" => {"key2" => nil}},
+    ]
+    target = build({
+                     type: :map,
+                     key: :string,
+                     item: :boolean,
+                   },
+                   values)
+    assert_equal(values, target.values)
+  end
```
Comment (kou) [inline ruby/red-arrow/test/values/test-sparse-union-array.rb:389]:
```suggestion
  end

```

File: ruby/red-arrow/lib/arrow/map-array-builder.rb (hunk 1 above)
Comment (kanga333) [reply ruby/red-arrow/lib/arrow/map-array-builder.rb]: Fixed 184b190

**Comment by kanga333** [comment]
@kou I've done the todo task and fixed the review points.
Could you please review this again?

File: ruby/red-arrow/test/values/test-sparse-union-array.rb [hunk 4]
```diff
@@ -373,6 +373,21 @@ def test_struct
...
+
+
```
Comment (kou) [inline ruby/red-arrow/test/values/test-sparse-union-array.rb:390]:
```suggestion
```

**Review by kou — APPROVED** [review]
+1

Thanks!

**Pull request closed without merging**